src/Microsoft.OData.Core/Evaluation/ODataResourceMetadataContext.cs

```diff
@@ -226,7 +226,37 @@ private static bool TryGetPrimitiveOrEnumPropertyValue(ODataResourceBase resourc
         {
             Debug.Assert(resource != null, "resource != null");
 
-            ODataProperty property = resource.NonComputedProperties == null ? null : resource.NonComputedProperties.SingleOrDefault(p => p.Name == propertyName);
+            ODataProperty property = null;
+            if (resource.NonComputedProperties != null)
+            {
+                // since this method is called frequently, we implement SingleOrDefault manually
+                // to avoid allocating predicate closures.
+                using (IEnumerator<ODataProperty> e = resource.NonComputedProperties.GetEnumerator())

```

Comment:
Why not do the check at the time of adding to `NonComputedProperties`? And/or use a data structure that can do this validation at point of adding?

Reply:
I think validating at the time of adding NonComputedProperties or using a better suited data structure would indeed be better. But I didn't want to risk breaking existing behaviour since I hadn't done a thorough analysis to ensure changing the behaviour here doesn't break any assumptions its callers.

Reply:
One challenge with validating the uniqueness properties at the time that they are added to `NonComputedProperties` is that we don't know when the properties are added to the collection. The `NonComputedProperties` refers to the `ODatarResource.Properties` property, which is an `IEnumerable<ODataProperty>`.

We can detect when the properties collection is set in the setter, and do the validation there, but we can't tell if new entries are added to the collection after the property has been set. For example, if we only validate the properties collection in the setter, we would not be able to catch the following violation:

```c#
var properties = new List<ODataProperty>
{
   new ODataProperty { Name = "Foo", Value = "Bar" },
   new ODataProperty { Name = "A", Value = "B" }
}

odataResource.Properties = properties; // duplicate name validation would occur here

properties.Add(new ODataProperty { Name = "Foo", Value = "Baz" }); // this duplicate property name violation would not be caught
```

It's worthwhile to note that we do already perform so other verification in the setter, and we can argue that the verification would also not catch violations that happen after the property has been set.

My only concern is that taking this route for the uniqueness check is effectively changing existing behaviour where a property is guaranteed to be unique at the type we query its value, even if the collection had changed since the `ODataResource.Properties` was set. This is behaviour if I have not yet verified is safe to break.

That said, my opinion is that we should probably leave it to the user to ensure they gave us unique properties and if they violate that we treat that as unsafe/undefined behaviour with unpredictable results.

It's also worth noting that the duplicate name checker will still verify property uniqueness independently of the check we do here, unless validation is disabled (in which case we probably shouldn't be doing that verification here either anyway). So, this check is sort of redundant.

Reply:
I've removed to the duplicate check and no unit test has failed. We also have other cases where we (questionably) use SingleOrDefault in our libraries, so maybe the original author did not added with intentions to ensure strong guarantees of uniqueness, but out of habit.

Reply:
I have revised the code and removed the duplicate check, the implementation now has `FirstOrDefault()` semantics instead of `SingleOrDefault()`.

I removed the duplicate-check from this method because:
- it doesn't break existing tests
- the duplicate check adds extra cost on a hot path

I opted **not** to move the duplicate-check to the `ODataResource.Properties` setter because:
- we already have a duplicate property name checking logic in the writer via `DuplicatePropertyNameChecker`
- in WebApi and OData Client, we can guarantee in the serializer that the property names are unique. That's also the best place to guarantee uniqueness since they have access the original property collection
- For customers using ODL directly, and who have disabled writer validation, we can assume they know what they're doing and already ensure that properties are unique.

However, if the last assumption doesn't hold, this could be a breaking change (though I think chances are extremely low). So, I'm open to restoring the original behaviour or moving the duplicate-check to the `ODataResource.Properties` setter if you or someone else believes it necessary to do so.

Reply:
Are the perf results still the same after removing the duplicate check?

Reply:
I had not collected alloc results after removing the duplicate check (I don't expect the result to be much different), but since I'm currently working on an area that calls this method, I've just taken CPU samples before and after applying this change:

You can see the `TryGetPrimitiveOrEnumPropertyValue` went down from 0.96% to 0.52% CPU. In the before graph, we can see that the lion's share was from `TryGetSingle` (which is called by `SingleOrDefault()`)

**Before**

![image](https://github.com/OData/odata.net/assets/8460169/ef1e8c8b-0881-4394-801a-38094fd62ff2)

**After**

![image](https://github.com/OData/odata.net/assets/8460169/3d1ef7b2-34bd-49e4-9bf5-a16765e97a1e)